core/src/main/scala/kafka/server/BrokerLifecycleManager.scala

```diff
@@ -507,6 +522,7 @@ class BrokerLifecycleManager(
         if (errorCode == Errors.NONE) {
           val responseData = message.data()
           failedAttempts = 0
+          offlineDirs = offlineDirs.map(kv => kv._1 -> true)
```

Comment:
I think this is incorrect. If a new failed directory is added to `offlineDirs` in-between a hearbeat request-resopnse, then we'll clear it here before knowing if it will propagated to the controller.

One idea is to hand down the offline dirs set in the request in `sendBrokerHeartBeat()` to `BrokerHeartbeatResponseEvent` through `BrokerHeartbeatResponseHandler` as a new constructor argument.